### PR TITLE
Add Postgres Adapter

### PR DESCRIPTION
**Pull request by vermakhushboo**
### Work in progress

**Edge cases in Appwrite but not in Utopia**

- Had to change the exception type to ‘Throwable’ instead of ‘PDOException’ for Appwrite as the exception wasn’t getting captured in createDocument and updateDocument methods
TODO: 
   - Add more tests that throw exceptions not caught by PDOException
   - Find places where we throw the wrong errors
   - Rethrow exceptions that Appwrite can understand

- There are not enough tests in Utopia to verify that double-quotes were needed around {$attribute} in the find() method to handle reserved words
TODO:
   - Find specific keywords that break this method
   - Automate iteration to test for keywords 
   - Ex: https://www.drupal.org/docs/develop/coding-standards/list-of-sql-reserved-words#s-reserved-words 
   - Find more reserved words for different databases, example, https://en.wikipedia.org/wiki/SQL_reserved_words

- Had to remove all alphanumeric values in getSQLCondition() method. Need to add better edge cases for full-text search in Utopia, it is missing many cases
TODO:
  - Add more tests to cover similar scenarios in Utopia 
  - For example email ids etc
  - Add tests specific to databases

**Improvements needed in Appwrite for future adapters**
TODO: 
- Hardcoded DB names in init.php, realtime.php, worker.php. Create generic interface for all adapters, allowing configurations using env variables
- Add health check method to each adapter
- Had to create a file PostgresSQLStatement.php and override Swoole methods, potentially move it to Utopia in the future

### CODE REVIEW

**Comment by KevinJ-hub** [comment]
Hey @vermakhushboo ,
I have started working on adding postgres adapter a few days back. I had created a [rfc](https://github.com/appwrite/rfc/pull/46) for the same and had a discussion regarding me starting to work on it on appwrites discord server as well.
Are you working on it too?

**Comment by vermakhushboo** [comment]
> Hey @vermakhushboo , I have started working on adding postgres adapter a few days back. I had created a [rfc](https://github.com/appwrite/rfc/pull/46) for the same and had a discussion regarding me starting to work on it on appwrites discord server as well. Are you working on it too?

Hey @KevinJ-hub, have you made considerable progress on the same? If yes, please take this forward, otherwise, I'd request you to pick something else.

**Comment by KevinJ-hub** [comment]
> @KevinJ-hub Please create a PR with your changes. The postgres adapter has a high priority at the moment and we want to ensure you have enough time to pick it up. We're expecting to ship it in the coming week. Let us know if you would be able to commit to it slightly_smiling_face

@christyjacob4 Ohhh, I won't be able to commit on finishing it in a week since I am currently a bit caught up with college work. If thats the case then its fine @vermakhushboo you can please continue with postgres adapter I might pick something else and work on it. 👍

**Comment by christyjacob4** [comment]
@KevinJ-hub Thank you so much for understanding! Really appreciate the honesty. If you need some guidance on issues you can pick up, please let me know. I'd be happy to assist.

File: src/Database/Adapter/Postgres.php [hunk 1]
```diff
@@ -1,129 +1,935 @@
...
+            $columns .= ":" . $column . ', ';
+        }
+
+        $stmt = $this->getPDO()
+            ->prepare("INSERT INTO \"{$this->getDefaultDatabase()}\".\"{$this->getNamespace()}_{$name}\"
+                ({$columnNames} _uid, _read, _write)
+                VALUES ({$columns} :_uid, :_read, :_write)");
+
+        $read = array_map(fn($role) => '"'.$role.'"', $document->getRead());
+        $write = array_map(fn($role) => '"'.$role.'"', $document->getWrite());
```
Comment (KevinJ-hub) [inline src/Database/Adapter/Postgres.php:346]: Hey @vermakhushboo,
The double quotes wrapped around role over here are redundent.

File: src/Database/Adapter/Postgres.php [hunk 2]
```diff
@@ -1,129 +1,935 @@
...
+            $column = $this->filter($attribute);
+            $columns .= "\"{$column}\"" . '=:' . $column . ',';
+        }
+
+        $stmt = $this->getPDO()
+            ->prepare("UPDATE \"{$this->getDefaultDatabase()}\".\"{$this->getNamespace()}_{$name}\"
+                SET {$columns} _uid = :_uid, _read = :_read, _write = :_write WHERE _uid = :_uid");
+
+        $read = array_map(fn($role) => '"'.$role.'"', $document->getRead());
+        $write = $read = array_map(fn($role) => '"'.$role.'"', $document->getWrite());
```
Comment (KevinJ-hub) [inline src/Database/Adapter/Postgres.php:413]: The same goes here too.

File: src/Database/Adapter/Postgres.php [hunk 3]
```diff
@@ -1,129 +1,935 @@
...
+            $columnNames .= "\"{$column}\"" . ', ';
+            $columns .= ":" . $column . ', ';
+        }
+
+        $stmt = $this->getPDO()
+            ->prepare("INSERT INTO \"{$this->getDefaultDatabase()}\".\"{$this->getNamespace()}_{$name}\"
+                ({$columnNames} _uid, _read, _write)
+                VALUES ({$columns} :_uid, :_read, :_write)");
+
+        $read = array_map(fn($role) => '"'.$role.'"', $document->getRead());
```
Comment (KevinJ-hub) [inline src/Database/Adapter/Postgres.php:345]: The double quotes wrapped around role seem to be redundent

File: src/Database/Adapter/Postgres.php (hunk 1 above)
Comment (KevinJ-hub) [inline src/Database/Adapter/Postgres.php:346]: Same as above

File: src/Database/Adapter/Postgres.php [hunk 4]
```diff
@@ -1,129 +1,935 @@
...
+        foreach ($attributes as $attribute => $value) { // Parse statement
+            $column = $this->filter($attribute);
+            $columns .= "\"{$column}\"" . '=:' . $column . ',';
+        }
+
+        $stmt = $this->getPDO()
+            ->prepare("UPDATE \"{$this->getDefaultDatabase()}\".\"{$this->getNamespace()}_{$name}\"
+                SET {$columns} _uid = :_uid, _read = :_read, _write = :_write WHERE _uid = :_uid");
+
+        $read = array_map(fn($role) => '"'.$role.'"', $document->getRead());
```
Comment (KevinJ-hub) [inline src/Database/Adapter/Postgres.php:412]: Same as above

File: src/Database/Adapter/Postgres.php (hunk 2 above)
Comment (KevinJ-hub) [inline src/Database/Adapter/Postgres.php:413]: Same as above

**Review by KevinJ-hub — CHANGES_REQUESTED** [review]
Hey @vermakhushboo ,
The double quotes around the role in _read and _write permission seem to be redundent over here since they do not get stored in the postgres DB. So I would like to request you to please check it once and remove them.

**Comment by eldadfux** [comment]
This PR is in an advance stage, and we hope to merge it in a few weeks. Right now we prioritize PRs that address DB reliability and usability before we add major new features.

File: src/Database/Adapter/Postgres.php [hunk 5]
```diff
@@ -1,129 +1,1202 @@
...
+        foreach ($attributes as $key => &$attribute) {
+            $length = $lengths[$key] ?? '';
+            $length = (empty($length)) ? '' : '(' . (int)$length . ')';
+            $order = $orders[$key] ?? '';
+            $attribute = $this->filter($attribute);
+
+            if (Database::INDEX_FULLTEXT === $type) {
+                $order = '';
+            }
+
```
Comment (fogelito) [inline src/Database/Adapter/Postgres.php:289]: I think we should  have the lower method for unique index strings something like because of case insensitive?
something like?
if(Database::INDEX_UNIQUE === $type) {
                $attribute = 'lower(' . $attribute . '::text)'; 
}

File: src/Database/Adapter/Postgres.php [hunk 6]
```diff
@@ -1,129 +1,1202 @@
...
+            if ($attribute->getAttribute('array')) {
+                $attrType = 'TEXT';
+            }
+
+            $attribute = "\"{$attrId}\" {$attrType}, ";
+        }
+
+        $stmt = $this->getPDO()
+            ->prepare("CREATE TABLE IF NOT EXISTS \"{$database}\".\"{$namespace}_{$id}\" (
```
Comment (abnegate) [inline src/Database/Adapter/Postgres.php:88]: Let's use the `getSQLTable` function here

File: src/Database/Adapter/Postgres.php [hunk 7]
```diff
@@ -1,129 +1,1202 @@
...
+        //INDEX (\"_createdAt\"),
+        //INDEX (\"_updatedAt\")
+        $stmtIndex = $this->getPDO()
+            ->prepare("CREATE UNIQUE INDEX \"index_{$namespace}_{$id}_uid\" on \"{$database}\".\"{$namespace}_{$id}\" (LOWER(_uid));");
+        try {
+            $stmt->execute();
+            $stmtIndex->execute();
+
+            $this->getPDO()
+                ->prepare("CREATE TABLE IF NOT EXISTS \"{$database}\".\"{$namespace}_{$id}_perms\" (
```
Comment (abnegate) [inline src/Database/Adapter/Postgres.php:107]: Let's use the `getSQLTable` function here

File: src/Database/Adapter/Postgres.php [hunk 8]
```diff
@@ -1,129 +1,1202 @@
...
+
+                if ($size >= 8) { // INT = 4 bytes, BIGINT = 8 bytes
+                    return 'BIGINT';
+                }
+
+                return 'INTEGER';
+                break;
+
+            case Database::VAR_FLOAT:
+                return 'REAL';
```
Comment (abnegate) [inline src/Database/Adapter/Postgres.php:1009]: Let's use `DOUBLE PRECISION` here

File: src/Database/Adapter/Postgres.php [hunk 9]
```diff
@@ -1,129 +1,1202 @@
...
+            case Database::INDEX_FULLTEXT:
+                $type = 'INDEX';
+                break;
+
+            default:
+                throw new Exception('Unknown Index Type:' . $type);
+                break;
+        }
+
+        return 'CREATE ' . $type . ' "' . $this->getNamespace() . '_' . $collection . '_' . $id . '" ON "' . $this->getDefaultDatabase() . '"."' . $this->getNamespace() . '_' . $collection . '" ( ' . implode(', ', $attributes) . ' );';
```
Comment (abnegate) [inline src/Database/Adapter/Postgres.php:1091]: Let's use `getSQLTable` here

**Review by abnegate — COMMENTED** [review]
Looks really good! One small type change and some readability updates and I think we'll be good to go :grin:

**Review by abnegate — APPROVED** [review]

**Review by fogelito — APPROVED** [review]
